http2: refactor Http2Stream and inbound headers

jasnell opened the pull request:
1. Eliminate pooling for `Nghttp2Stream` instances
2. Use a MaybeStackBuffer for inbound headers
3. Enforce a maximum number of headers per block
4. Enforce a maximum number of header octets per block using MAX_HEADERS_LIST setting

##### Checklist
- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
http2

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/11141/

mcollina commented on doc/api/http2.md:
```diff
@@ -1773,8 +1779,8 @@ properties.
-  of header list that will be accepted. There is no default value. The minimum
-  allowed value is 0. The maximum allowed value is 2<sup>32</sup>-1.
+  of header list that will be accepted. The minimum allowed value is 0. The
+  maximum allowed value is 2<sup>32</sup>-1. **Default:** 65535.
```
we should probably update the changelog in the docs somehow for this.

jasnell replied:
Good point

mcollina commented on src/node_http2_core-inl.h:
```diff
@@ -630,44 +647,31 @@ inline void Nghttp2Session::RemoveStream(int32_t id) {
+  uint32_t maxHeaderListSize = session->GetMaxHeaderPairs();
+  if (maxHeaderListSize == 0)
+    maxHeaderListSize = DEFAULT_MAX_HEADER_LIST_PAIRS;
+  current_headers_.AllocateSufficientStorage(maxHeaderListSize);
```
does this mean we are preallocating an array of `maxHeaderListSize`? Do we need this to be 65535 as the default? IMHO we are wasting memory and time for each stream.

jasnell replied:
There are two limits: the max number of header pairs, and the max number of header octets. We are preallocating an array of nhgttp2_header structs. These consist of two pointers and an 8 bit flag. Current number is a generous 1024, which can likely be tuned much much lower. The 65535 is the total number of octets and is maintained as a size_t counter only. There's no preallocation for that.

We should be able to tune the number of header pairs down to something much much smaller than 1024.

mcollina replied:
I would go for 128

addaleax replied:
If you’re calling `AllocateSufficientStorage()`, you kind of shouldn’t be using a `MaybeStackBuffer` in the first place – all this will give you is unused memory inside the `Nghttp2Stream` instance.

I’d suggest just going with a `std::vector` here … you don’t need to care about pre-allocating memory in that case, plus you can *still* reserve memory ahead of time using e.g. `vector.reserve(128)`.

jasnell replied:
yeah, I keep going back and forth on this, just playing with the various options to see which is going to provide the best performance. `std::vector` is likely the best approach based on everything I've seen.

mcollina commented on test/parallel/test-http2-too-many-headers.js:
```diff
@@ -0,0 +1,34 @@
+// block is 1024, including the HTTP pseudo-header fields, by
+// setting one here, no request should be acccepted because it
+// does not allow for even the basic HTTP pseudo-headers
+const server = http2.createServer({ maxHeaderListPairs: 1 });
```
I think these types of settings should throw. What's the minimum?

jasnell replied:
We can set whatever minimum we want here. For the request to work, we need at least 4 to account for the required pseudoheaders.

mcollina replied:
I would say we validate it's greater than 4.

jasnell replied:
how about rather than report an error, we just automatically allow for the number of pseudo-headers? So if a user sets 0, the internal value is 4....

addaleax commented on src/node_http2.cc:
```diff
@@ -258,11 +260,14 @@ void RefreshDefaultSettings(const FunctionCallbackInfo<Value>& args) {
     (1 << IDX_SETTINGS_INITIAL_WINDOW_SIZE) |
-    (1 << IDX_SETTINGS_MAX_FRAME_SIZE);
+    (1 << IDX_SETTINGS_MAX_FRAME_SIZE) |
+    (1 << IDX_SETTINGS_MAX_HEADER_LIST_SIZE);
```
btw, this is … clever :smile:

addaleax commented on src/node_http2.cc:
```diff
@@ -975,18 +981,19 @@ void Http2Session::OnHeaders(
-    while (!headers->empty() && j < arraysize(argv) / 2) {
-      nghttp2_header item = headers->front();
+    while (count > 0 && j < arraysize(argv) / 2) {
+      nghttp2_header item = headers[n++];
```
Can’t this just be `headers[--count]`? I don’t think you really need the `n` if you didn’t need it before

jasnell commented:
@mcollina @addaleax ... updated.

1. Set the default header pairs limit to 128
2. Updated so that on the server, 4 is the minimum. On the client, 1 is the minimum.
3. Updated to use a std::vector

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/11197/

jasnell commented:
Landed in 9f3d59e
